### PR TITLE
Feat/is 399 ggs create file

### PR DESCRIPTION
**Pull request by alexanderleegs**
## Problem

This PR introduces creates to our GGS solution. Resolves IS-399. This PR also modifies checks for the existence of paths to use `gitFilePathStats` instead of `fs.existsSync`.

**Features**:

- Refactored safeExistsSync calls in `clone` and `isValidGitRepo` to use `getFilePathStats` instead
- Create operation supported for GGS

Future work:

- Verify image creation - pending image retrieval PR

### CODE REVIEW

File: src/services/db/GitFileSystemService.ts [hunk 1]
```diff
@@ -433,9 +427,104 @@ export default class GitFileSystemService {
...
-  // ResourceDirectoryService used this to create a directory + file at the same time
-  create() {}
+  // Creates a file and the associated directory if it doesn't exist
+  create({
+    repoName,
+    userId,
+    content,
+    directoryName,
+    fileName,
+  }: {
```
Comment (dcshzj) [inline src/services/db/GitFileSystemService.ts:437]: Should we standardise to passing function arguments as-is? Or should we pass as object instead like this?

File: src/services/db/GitFileSystemService.ts [hunk 2]
```diff
@@ -433,9 +427,104 @@ export default class GitFileSystemService {
...
+    repoName: string
+    userId: string
+    content: string
+    directoryName: string
+    fileName: string
+  }): ResultAsync<
+    { sha: string },
```
Comment (dcshzj) [inline src/services/db/GitFileSystemService.ts:444]: This should be `GitCommitResult`.

File: src/services/db/GitFileSystemService.ts [hunk 3]
```diff
@@ -433,9 +427,104 @@ export default class GitFileSystemService {
...
+        }
+        return err(error)
+      })
+      .andThen(() => this.getFilePathStats(repoName, filePath))
+      .andThen((stats) => {
+        if (stats.isFile())
+          return err(
+            new ConflictError(
+              `File ${filePath} already exists in repo ${repoName}`
+            )
```
Comment (dcshzj) [inline src/services/db/GitFileSystemService.ts:485]: I think we might potentially need to rollback the earlier creation of the directory if it didn't exist originally.

File: src/services/db/GitFileSystemService.ts [hunk 4]
```diff
@@ -433,9 +427,104 @@ export default class GitFileSystemService {
...
+          (error) => {
+            logger.error(`Error when creating ${filePath}: ${error}`)
+            if (error instanceof Error) {
+              return new GitFileSystemError(error.message)
+            }
+            return new GitFileSystemError("An unknown error occurred")
+          }
+        )
+          .map((res) => res)
+          .mapErr((error) => error)
```
Comment (dcshzj) [inline src/services/db/GitFileSystemService.ts:507]: Hmm can I check what does this part do?

File: src/services/db/GitFileSystemService.ts [hunk 5]
```diff
@@ -433,9 +427,104 @@ export default class GitFileSystemService {
...
+      })
+      .andThen(() =>
+        ResultAsync.fromPromise(
+          fs.promises.writeFile(pathToEfsFile, encodedContent),
+          (error) => {
+            logger.error(`Error when creating ${filePath}: ${error}`)
+            if (error instanceof Error) {
+              return new GitFileSystemError(error.message)
+            }
+            return new GitFileSystemError("An unknown error occurred")
```
Comment (dcshzj) [inline src/services/db/GitFileSystemService.ts:503]: We potentially might need a rollback here.

File: src/services/db/RepoService.ts [hunk 6]
```diff
@@ -113,9 +113,36 @@ export default class RepoService extends GitHubService {
...
+      fileName,
+      directoryName,
+      isMedia = false,
+    }: {
+      content: string
+      fileName: string
+      directoryName: string
+      isMedia?: boolean
+    }
+  ): Promise<{ sha: string }> {
```
Comment (dcshzj) [inline src/services/db/RepoService.ts:128]: Same here, this should be `GitCommitResult`.

File: src/services/db/GitFileSystemService.ts (hunk 3 above)
Comment (alexanderleegs) [reply src/services/db/GitFileSystemService.ts:485]: If the file already exists, the directory definitely exists, so no need to rollback! Also github doesn't care about directory structure, you can only specify directory structure by pushing files into that directory, so this wouldn't affect us regardless!

Comment (dcshzj) [reply src/services/db/GitFileSystemService.ts:485]: Oops yup, you're right! Thanks for clarifying!

File: src/services/db/GitFileSystemService.ts [hunk 7]
```diff
@@ -1,5 +1,6 @@
 import fs from "fs"
 
+import { Base64 } from "js-base64"
```
Comment (dcshzj) [inline src/services/db/GitFileSystemService.ts:3]: Oh yeah, do we still need this?

File: src/services/db/RepoService.ts (hunk 6 above)
Comment (alexanderleegs) [reply src/services/db/RepoService.ts:128]: The result here needs to be `{ sha: string }` in order to match the result of the `githubservice.create`!

Comment (dcshzj) [reply src/services/db/RepoService.ts:128]: Oh gosh 😭

File: src/services/db/GitFileSystemService.ts (hunk 1 above)
Comment (alexanderleegs) [reply src/services/db/GitFileSystemService.ts]: Hmm i think it might be clearer to pass as an object, but let's swap it to passing as in for consistency! Done in https://github.com/isomerpages/isomercms-backend/pull/892/commits/c6190b92eddd62f95ede7100a207565ed8a9e555

File: src/services/db/GitFileSystemService.ts (hunk 2 above)
Comment (alexanderleegs) [reply src/services/db/GitFileSystemService.ts]: Changed in https://github.com/isomerpages/isomercms-backend/pull/892/commits/b0b811aec0ce097f1595fbfd772bb64d271c09da

File: src/services/db/GitFileSystemService.ts (hunk 4 above)
Comment (alexanderleegs) [reply src/services/db/GitFileSystemService.ts]: Whoops this isn't needed anymore! https://github.com/isomerpages/isomercms-backend/pull/892/commits/b8b22f19298a33df4d10ade42f05416a0d5cfcb6

File: src/services/db/GitFileSystemService.ts (hunk 5 above)
Comment (alexanderleegs) [reply src/services/db/GitFileSystemService.ts]: Fixed in https://github.com/isomerpages/isomercms-backend/pull/892/commits/a1dd5c2b37fc75410e1f58d9c9c9d4a4f2cde9a4

File: src/services/db/GitFileSystemService.ts (hunk 7 above)
Comment (alexanderleegs) [reply src/services/db/GitFileSystemService.ts]: Not in this PR at least, it might be needed for images? But if work needs to be done on that aspect we can tackle it in a separate PR! Removed in https://github.com/isomerpages/isomercms-backend/pull/892/commits/a4776f561cbe8392a1f208d940543c6406b867c3

File: src/services/db/RepoService.ts (hunk 6 above)
Comment (dcshzj) [reply src/services/db/RepoService.ts:128]: I think we can leave it as it is right now, but I think we can potentially refactor in the future so that they have the same return types

File: src/services/configServices/NetlifyTomlService.js [hunk 8]
```diff
@@ -1,3 +1,4 @@
+const { Base64 } = require("js-base64")
```
Comment (dcshzj) [inline src/services/configServices/NetlifyTomlService.js:1]: Hmm this one also :(

**Review by dcshzj — APPROVED** [review]

File: src/services/configServices/NetlifyTomlService.js (hunk 8 above)
Comment (alexanderleegs) [reply src/services/configServices/NetlifyTomlService.js:1]: This one is needed! For some reason it wasn't being imported before

File: src/services/db/GitFileSystemService.ts [hunk 9]
```diff
@@ -82,26 +86,22 @@ export default class GitFileSystemService {
...
-      if (error instanceof GitError || error instanceof Error) {
-        return new GitFileSystemError("Unable to determine if directory exists")
-      }
-
-      return new GitFileSystemError("An unknown error occurred")
-    })
-
-    return safeExistsSync(`${EFS_VOL_PATH}/${repoName}`)
-      .andThen((isFolderExisting) => {
-        if (!isFolderExisting) {
```
Comment (harishv7) [inline src/services/db/GitFileSystemService.ts:97]: why are these removed?

File: src/services/db/GitFileSystemService.ts [hunk 10]
```diff
@@ -228,19 +228,17 @@ export default class GitFileSystemService {
...
-      if (error instanceof GitError || error instanceof Error) {
-        return new GitFileSystemError("Unable to determine if directory exists")
-      }
-
-      return new GitFileSystemError("An unknown error occurred")
-    })
-
-    return safeExistsSync(`${EFS_VOL_PATH}/${repoName}`).asyncAndThen(
-      (isFolderExisting) => {
-        if (!isFolderExisting) {
```
Comment (harishv7) [inline src/services/db/GitFileSystemService.ts:243]: same here, do we not need these anymore?

File: src/services/db/GitFileSystemService.ts [hunk 11]
```diff
@@ -433,9 +430,98 @@ export default class GitFileSystemService {
...
+      .andThen((latestCommit) => {
+        const { sha } = latestCommit
+        if (!sha) {
+          return errAsync(new GitFileSystemError("An unknown error occurred"))
+        }
+        oldStateSha = sha
+        return okAsync(true)
+      })
+      .andThen(() => this.getFilePathStats(repoName, directoryName))
+      .andThen(() => ok(""))
```
Comment (harishv7) [inline src/services/db/GitFileSystemService.ts:460]: what is this empty string for, seems like a redundant step?

File: src/services/db/GitFileSystemService.ts [hunk 12]
```diff
@@ -433,9 +430,98 @@ export default class GitFileSystemService {
...
-  // TODO: Creates either directory or file
-  // ResourceDirectoryService used this to create a directory + file at the same time
-  create() {}
+  // Creates a file and the associated directory if it doesn't exist
+  create(
+    repoName: string,
+    userId: string,
+    content: string,
+    directoryName: string,
+    fileName: string
```
Comment (harishv7) [inline src/services/db/GitFileSystemService.ts:439]: how do we differentiate between a file creation vs directory creation? should we add an explicit isDir flag same as delete?

File: src/services/db/GitFileSystemService.ts [hunk 13]
```diff
@@ -433,9 +430,98 @@ export default class GitFileSystemService {
...
+        if (stats.isFile())
+          return err(
+            new ConflictError(
+              `File ${filePath} already exists in repo ${repoName}`
+            )
+          )
+        return ok("")
+      })
+      .orElse((error) => {
+        if (error instanceof NotFoundError) {
```
Comment (harishv7) [inline src/services/db/GitFileSystemService.ts:487]: where do we get the NotFoundError from?

File: src/services/db/GitFileSystemService.ts [hunk 14]
```diff
@@ -433,9 +430,98 @@ export default class GitFileSystemService {
...
+      })
+      .andThen(() => this.getFilePathStats(repoName, filePath))
+      .andThen((stats) => {
+        if (stats.isFile())
+          return err(
+            new ConflictError(
+              `File ${filePath} already exists in repo ${repoName}`
+            )
+          )
+        return ok("")
```
Comment (harishv7) [inline src/services/db/GitFileSystemService.ts:484]: nit: maybe can use boolean ok(true)

File: src/services/db/RepoService.ts [hunk 15]
```diff
@@ -113,9 +113,36 @@ export default class RepoService extends GitHubService {
...
+    if (this.isRepoWhitelisted(sessionData.siteName)) {
+      logger.info("Writing file to local Git file system")
```
Comment (harishv7) [inline src/services/db/RepoService.ts:130]: can we log for which repoName is this along with the dirName and fileName?

File: src/services/db/RepoService.ts [hunk 16]
```diff
@@ -113,9 +113,36 @@ export default class RepoService extends GitHubService {
...
-  ): Promise<any> {
+    sessionData: UserWithSiteSessionData,
+    {
+      content,
+      fileName,
+      directoryName,
+      isMedia = false,
+    }: {
+      content: string
+      fileName: string
```
Comment (harishv7) [inline src/services/db/RepoService.ts:124]: will filename be there for create dir too? if not, is this optional? 
some thinking points - would it be better to have 2 separate functions for createFile createDir or maybe a flag like isDir to make it explicit?

File: src/services/db/__tests__/GitFileSystemService.spec.ts [hunk 17]
```diff
@@ -57,6 +58,140 @@ describe("GitFileSystemService", () => {
...
+        size: "Another fake content".length,
+      }
+
+      const result = await GitFileSystemService.listDirectoryContents(
+        "fake-repo",
+        ""
+      )
+      const actual = result
+        ._unsafeUnwrap()
+        .sort((a, b) => a.name.localeCompare(b.name))
```
Comment (harishv7) [inline src/services/db/__tests__/GitFileSystemService.spec.ts:101]: why is this sort important?

File: src/services/db/__tests__/GitFileSystemService.spec.ts [hunk 18]
```diff
@@ -677,182 +812,224 @@ describe("GitFileSystemService", () => {
...
         "fake-repo",
-        "fake-dir/fake-file"
+        "fake-user-id",
+        "fake content",
+        "fake-dir",
+        "fake-file"
       )
 
-      expect(result.isErr()).toBeTrue()
+      expect(actual.isErr()).toBeTrue()
```
Comment (harishv7) [inline src/services/db/__tests__/GitFileSystemService.spec.ts:935]: we can assert the type of error returned with instanceOf

File: src/services/db/__tests__/GitFileSystemService.spec.ts [hunk 19]
```diff
@@ -677,182 +812,224 @@ describe("GitFileSystemService", () => {
...
+      })
+
+      const result = await GitFileSystemService.read(
         "fake-repo",
-        "non-existent-dir"
+        "fake-dir/fake-file"
       )
 
-      expect(result._unsafeUnwrapErr()).toBeInstanceOf(NotFoundError)
+      expect(result.isErr()).toBeTrue()
```
Comment (harishv7) [inline src/services/db/__tests__/GitFileSystemService.spec.ts:1032]: how about tests to ensure rollbacks are done correctly - i.e. when failures occur along one of the many steps in create flow

File: src/services/db/GitFileSystemService.ts (hunk 9 above)
Comment (alexanderleegs) [reply src/services/db/GitFileSystemService.ts:97]: We've opted to use `gitFilePathStats` instead of `fs.existsSync`, so these have been removed!

File: src/services/db/GitFileSystemService.ts (hunk 12 above)
Comment (alexanderleegs) [reply src/services/db/GitFileSystemService.ts:439]: We never explicitly create a directory within our existing codebase (since github only takes the presence of files within a directory to imply the existence of the directory - previously we'd just create the file directly). Previous "directory creation" operations were actually just creating a `.keep` file in that particular directory path, to ensure it showed up in github, and this behaviour stays the same, so there's no need to have a differentiated argument here

File: src/services/db/GitFileSystemService.ts (hunk 13 above)
Comment (alexanderleegs) [reply src/services/db/GitFileSystemService.ts:487]: This is from `this.getFilePathStats(repoName, filePath)` - if it's not found, the file does not yet exist

File: src/services/db/RepoService.ts (hunk 16 above)
Comment (alexanderleegs) [reply src/services/db/RepoService.ts:124]: Yup it'll be there as well, create dir operations are just creating a `.keep` file in the desired directory currently

File: src/services/db/__tests__/GitFileSystemService.spec.ts (hunk 17 above)
Comment (alexanderleegs) [reply src/services/db/__tests__/GitFileSystemService.spec.ts:101]: This is the existing test behaviour - i moved the position of this test because it was failing for some reason haha

I think the sort is to ensure that the retrieval always returns in the same order, regardless of how it's returned

Comment (dcshzj) [reply src/services/db/__tests__/GitFileSystemService.spec.ts:101]: Yup, to add on, jest also compares the order, even if the items are the same. This would ensure that the unit test does not get flaky for no reason.

File: src/services/db/__tests__/GitFileSystemService.spec.ts (hunk 19 above)
Comment (alexanderleegs) [reply src/services/db/__tests__/GitFileSystemService.spec.ts:1032]: Hmm we actually only rollback in either the actual file creation, or when committing the change. I'm not entirely sure how to mock the fs call itself!

File: src/services/db/GitFileSystemService.ts (hunk 10 above)
Comment (alexanderleegs) [reply src/services/db/GitFileSystemService.ts]: See above!

File: src/services/db/GitFileSystemService.ts (hunk 11 above)
Comment (alexanderleegs) [reply src/services/db/GitFileSystemService.ts]: Whoops it's meant to convert stats into a boolean value but the functionality wasn't implemented fully, this has been modified in https://github.com/isomerpages/isomercms-backend/pull/892/commits/b3d87e344ceb2d146ceaf661b40c9548dcc5102d

File: src/services/db/GitFileSystemService.ts (hunk 14 above)
Comment (alexanderleegs) [reply src/services/db/GitFileSystemService.ts]: Changed in https://github.com/isomerpages/isomercms-backend/pull/892/commits/b3d87e344ceb2d146ceaf661b40c9548dcc5102d!

File: src/services/db/__tests__/GitFileSystemService.spec.ts (hunk 17 above)
Comment (harishv7) [reply src/services/db/__tests__/GitFileSystemService.spec.ts:101]: I see, I believe there is a jest function like strict equals or something to remove impact of order. But this looks good to me!

File: src/services/db/RepoService.ts (hunk 15 above)
Comment (alexanderleegs) [reply src/services/db/RepoService.ts]: https://github.com/isomerpages/isomercms-backend/pull/892/commits/20d3539fadd5f6e1abbe84a363637c69a9172a19

File: src/services/db/__tests__/GitFileSystemService.spec.ts (hunk 18 above)
Comment (alexanderleegs) [reply src/services/db/__tests__/GitFileSystemService.spec.ts]: Fixed in https://github.com/isomerpages/isomercms-backend/pull/892/commits/f0dd458b5ecea65092b685a34ec2e332d591f997

File: src/services/db/GitFileSystemService.ts (hunk 12 above)
Comment (harishv7) [reply src/services/db/GitFileSystemService.ts:442]: i thought GitFileSystemService will decide to create the .keep but actually it comes from the upstream directory services

**Review by harishv7 — APPROVED** [review]

File: src/services/db/__tests__/GitFileSystemService.spec.ts (hunk 19 above)
Comment (harishv7) [reply src/services/db/__tests__/GitFileSystemService.spec.ts:1032]: okay, let's focus on the manual testing for rollback and add this in later. I think @dcshzj added a mock-fs library into the spec file which we can use possibly
